OmsAgent/watcherutil.py

```diff
@@ -162,3 +251,261 @@ def watch(self):
+                take_action = True 
+
+            if (take_action):
+                if (self._consecutive_restarts_due_to_error < 5):
+                    self.take_corrective_action(self_mon_info)                                    
+                    self._consecutive_restarts_due_to_error += 1
+                else:
+                    self._hutil_error("Last 5 restarts did not help. So we will not restart the agent immediately")
+                    
+                    # Reset historical infomration. 
+                    self._consecutive_restarts_due_to_error = 0    
+                    self_mon_info.reset_error_info()
+        else:
+            """
+                If we are able to get the heartbeats, check omsagent logs
+                to identify if there are any error logs.
+            """                                                                    
+            self_mon_info.reset_error_info()   
+            self._consecutive_restarts_due_to_error = 0       
+
+    def monitor_resource(self, self_mon_info):
+        """
+            Monitor resource utilization of omsagent.
+            Check for memory and CPU periodically. If they cross the threshold for consecutive 3 iterations
+            we will restart the agent. 
+        """
+
+        resource_usage = self.get_oms_agent_resource_usage()
+        message = "Memory : {0}, CPU : {1}".format(resource_usage[0], resource_usage[1])
+        event = self.create_telemetry_event("agenttelemetry","True",message,"300000")        
+        self.write_waagent_event(event)
+
+        self_mon_info._memory_used_in_percent = resource_usage[0]
+                
+        if (self_mon_info._memory_used_in_percent > 0):
+            if (self_mon_info._memory_used_in_percent > MemoryThresholdToWatchFor):
+                # check consecutive memory usage.
+                self_mon_info.increment_high_memory_count()
+                if (self_mon_info.corssed_memory_threshold()):
+                    # if we have crossed the memory threshold take corrective action.
+                    self.take_corrective_action(self_mon_info)
+                else:
+                    self_mon_info.reset_high_memory_count()
+            else:
+                self_mon_info.reset_high_memory_count()
+
+    def monitor_health(self):
+        """
+            Role of this function is monitor the health of the oms agent.
+            To begin with it will monitor heartbeats flowing through oms agent. 
+            We will also read oms agent logs to determine some error conditions.
+            We don't want to interfare with log watcher function.
+            So we will start this on a new thread.
+        """
+
+        self_mon_info = SelfMonitorInfo()
+        log_file_marker = LogFileMarker()
+        
+        # check every 6 minutes. we want to be bit pessimistic while looking for health, especially heartbeats which is emitted every 5 minutes.
+        sleepTime =  6 * 60  
```

Comment:
A 15 minute interval might be better here.

Reply:
Heartbeat is sent every 5 minutes for sure with buffering it can take additional few seconds so 6 minutes should be more than enough. also it is better to catch the issues sooner than later.